CONTRIBUTING.md

```diff
@@ -78,6 +78,123 @@ available in bootstrap classloader.
 Gradle plugin.
 
+### Writing instrumentation
+
+**Warning**: The repository is still in the process of migrating to the structure described here.
+
+Any time we want to add OpenTelemetry support for a new Java library, e.g., so usage
+of that library has tracing, we must write new instrumentation for that library. Let's
+go over some terms first.
+
+**Manual Instrumentation**: This is logic that creates spans and enriches them with data
+using library-specific monitoring APIs. For example, when instrumenting an RPC library,
+the instrumentation will use some library-specific functionality to listen to events such
+as the start and end of a request and will execute code to start and end spans in these
+listeners. Many of these libraries will provide interception type APIs such as the gRPC
+`ClientInterceptor` or servlet's `Filter`. Others will provide a Java interface whose methods
+correspond to a request, and instrumentation can define an implementation which delegates
+to the standard, wrapping methods with the logic to manage spans. Users will add code to their
+apps that initialize the classes provided by manual instrumentation libraries and the libraries
+can be found inside the user's app itself.
+
+Some libraries will have no way of intercepting requests because they only expose static APIs
+and no interception hooks. For these libraries it is not possible to create manual
+instrumentation.
+
+**Auto Instrumentation**: This is logic that is similar to manual instrumentation, but instead
+of a user initializing classes themselves, a Java agent automatically initializes them during
+class loading by manipulating byte code. This allows a user to develop their apps without thinking
+about instrumentation and get it "for free". Often, the auto instrumentation will generate bytecode
+that is more or less identical to what a user would have written themselves in their app.
+
+In addition to automatically initializing manual instrumentation, auto instrumentation can be used
+for libraries where manual instrumentation is not possible, such as `URLConnection`, because it can
+intercept even the JDK's classes. Such libraries will not have manual instrumentation but will have
+auto instrumentation.
+
+#### Folder Structure
+
+When writing new instrumentation, create a new subfolder of `instrumentation` to correspond to the
+instrumented library and the version being targeted. Ideally an old version of the library is
+targeted in a way that the instrumentation applies to a large range of versions, but this may be
+restricted by the interception APIs provided by the library.
+
+Within the subfolder, create three folders `library` (skip if manual instrumentation is not possible),
+`agent`, and `testing`.
+
+For example, if we are targeting an RPC framework `yarpc` at version `1.0` we would have a tree like
+
+```
+instrumentation ->
+    ...
+    yarpc-1.0 ->
+        agent
+            agent.gradle
```

Comment:
Gradle files names should be unique, so likely `yarpc-1.0-agent.gradle`

Reply:
Maybe @iNikem meant this?

```
instrumentation ->
  yarpc ->
    yarpc-1.0 ->
      agent ->
        yarpc-1.0-agent.gradle
      library ->
        yarpc-1.0-library.gradle
```

and module names
* `:yarpc:yarpc-1.0:agent`
* `:yarpc:yarpc-1.0:library`

Reply:
I think with that, we would need something like this in settings.gradle

```
project(':instrumentation:yarpc:yarpc-1.0:agent').name = 'yarpc-1.0-agent'
```

which is pretty tedious. Don't believe there is a workaround for that since it's at the initial loading of the projects.

Reply:
not saying it's a good idea, but we could probably add that logic to this existing code in `settings.gradle`:

```
def setBuildFile(project) {
  project.buildFileName = "${project.name}.gradle"
  project.children.each {
    setBuildFile(it)
  }
}

rootProject.children.each {
  setBuildFile(it)
}
```

Reply:
I never noticed this code I could indeed tweak it. @inikem is @trask's suggested structure what you were thinking?

Reply:
Yes

Reply:
It doesn't bother me, but I'm not well-versed in gradle convention, so nothing stands out to me as particularly off.

Just throwing out another option if you want to keep gradle file in sync with module name:

```
instrumentation ->
  yarpc ->
    v1.0 ->    (i'm just guessing using only 1.0 without the "v" prefix might be an issue?)
      yarpc-1.0-agent ->
        yarpc-1.0-agent.gradle
      yarpc-1.0-library ->
        yarpc-1.0-library.gradle
    v2.0 ->
      yarpc-2.0-agent ->
        yarpc-2.0-agent.gradle
      yarpc-2.0-library ->
        yarpc-2.0-library.gradle
```

Reply:
I don't feel too strongly either, if both of you are ok with it than I'm ok with the current form too :)